### PR TITLE
feat: listing duplication endpoint

### PR DESCRIPTION
**Pull request by mcgarrye**
This PR addresses [#(799)](https://app.zenhub.com/workspaces/bloom-5dc32d7144bd400001315dac/issues/gh/metrotranscom/doorway/799)

- [x] Addresses the issue in full
- [ ] Addresses only certain aspects of the issue

## Description

Allows user to call `listings/duplicate` endpoint to create a copy listing from an existing listing.

## How Can This Be Tested/Reviewed?

Run partner portal and api and login as a admin user.
Have the listing id for the listing you would like to duplicate.
Fill out the payload for `listings/duplicate`. Set storedListing.id to the above.
Cases:

1. Set name to a unique name, set includeUnits to true. Should create a new draft listing with the name given which is otherwise identical to the original listing, including the units.
2. Set name to a unique name, set includeUnits to false. Should create a new draft listing with the name given which is otherwise identical to the original listing, excluding the units.
3. Set name to a the same name as the original listing. Should error with 400 BadRequest and a message 'New listing name must be unique'.

You can view the newly created in the Partners Portal

## Author Checklist:

- [x] Added QA notes to the issue with applicable URLs
- [x] Reviewed in a desktop view
- [ ] Reviewed in a mobile view
- [ ] Reviewed considering accessibility
- [x] Added tests covering the changes
- [ ] Made corresponding changes to the documentation
- [x] Ran `yarn generate:client` and/or created a migration when required

## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates

### CODE REVIEW

File: api/src/services/listing.service.ts [hunk 1]
```diff
@@ -965,6 +967,68 @@ export class ListingService implements OnModuleInit {
...
+      delete mappedListing['units'];
+    }
+
+    const newListingData: ListingCreate = {
+      ...mappedListing,
+      name: dto.name,
+      status: ListingsStatusEnum.pending,
+      listingEvents: listingEvents,
+      listingMultiselectQuestions:
+        storedListing.listingMultiselectQuestions?.map((question) => {
```
Comment (ColinBuyck) [inline api/src/services/listing.service.ts:1018]: Nit: Could we use that =>({}) syntax brought up in yesterdays huddle?

File: api/src/services/listing.service.ts [hunk 2]
```diff
@@ -965,6 +967,68 @@ export class ListingService implements OnModuleInit {
...
+          return {
+            id: question.multiselectQuestionId,
+            ordinal: question.ordinal,
+          };
+        }),
+      listingImages: listingImages,
+      lotteryLastRunAt: undefined,
```
Comment (ColinBuyck) [inline api/src/services/listing.service.ts:1025]: Just merged in a PR adding lotteryLastPublishedAt field which should be set to undefined here too

File: api/test/integration/listing.e2e-spec.ts [hunk 3]
```diff
@@ -661,6 +661,57 @@ describe('Listing Controller Tests', () => {
     });
   });
 
+  describe('duplicate endpoint', () => {
```
Comment (ColinBuyck) [inline api/test/integration/listing.e2e-spec.ts:664]: Nit: Can these be separated to improve readability on if both duplication calls are broken or if it is dependent on the units piece?

**Review by ColinBuyck — APPROVED** [review]
LGTM 🫕 Just a few nits

File: api/.env.template [hunk 4]
```diff
@@ -44,6 +44,8 @@ LOTTERY_PUBLISH_PROCESSING_CRON_STRING=58 23 * * *
 LOTTERY_PROCESSING_CRON_STRING=0 * * * *
 # how many days till lottery data expires
 LOTTERY_DAYS_TILL_EXPIRY=45
+# allow partners and jadmins to create duplicate listings
+ALLOW_PARTNERS_TO_CREATE_DUPLICATES=FALSE
```
Comment (YazeedLoonat) [inline api/.env.template:48]: can we change this to `ALLOW_PARTNERS_TO_DUPE_LISTINGS` or something like that, to make it clear that this is about duplicating listings

File: api/src/services/listing.service.ts [hunk 5]
```diff
@@ -965,6 +967,105 @@ export class ListingService implements OnModuleInit {
...
+    await this.permissionService.canOrThrow(
+      { ...requestingUser, userRoles: userRoles },
+      'listing',
+      permissionActions.create,
+      {
+        jurisdictionId: storedListing.jurisdictions.id,
+      },
+    );
+
+    if (dto.name.trim() === storedListing.name) {
```
Comment (YazeedLoonat) [inline api/src/services/listing.service.ts:998]: can you move this to right after we get the stored listing, so we error early and don't do more work than we need to

**Review by YazeedLoonat — CHANGES_REQUESTED** [review]
looks good, some changes requested!

**Review by ColinBuyck — APPROVED** [review]
Yahoooo 🪨

**Pull request merged**